### PR TITLE
Infra Policy Mapping

### PR DESCRIPTION
**Pull request by Mykhailenko**

### CODE REVIEW

File: config/rm/nodesource/infrastructures-policies-mapping [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+org.ow2.proactive.resourcemanager.nodesource.infrastructure.LSFInfrastructure,org.ow2.proactive.resourcemanager.nodesource.policy.StaticPolicy,org.ow2.proactive.resourcemanager.nodesource.policy.TimeSlotPolicy,org.ow2.proactive.scheduler.resourcemanager.nodesource.policy.ReleaseResourcesWhenSchedulerIdle,org.ow2.proactive.resourcemanager.nodesource.policy.CronPolicy,org.ow2.proactive.scheduler.resourcemanager.nodesource.policy.CronSlotLoadBasedPolicy
+org.ow2.proactive.resourcemanager.nodesource.infrastructure.DefaultInfrastructureManager,org.ow2.proactive.resourcemanager.nodesource.policy.StaticPolicy,org.ow2.proactive.resourcemanager.nodesource.policy.TimeSlotPolicy,org.ow2.proactive.scheduler.resourcemanager.nodesource.policy.ReleaseResourcesWhenSchedulerIdle,org.ow2.proactive.resourcemanager.nodesource.policy.CronPolicy,org.ow2.proactive.scheduler.resourcemanager.nodesource.policy.CronSlotLoadBasedPolicy
+
+org.ow2.proactive.resourcemanager.nodesource.infrastructure.NativeSchedulerInfrastructure,org.ow2.proactive.scheduler.resourcemanager.nodesource.policy.NativeSchedulerPolicy
+
+
+
```
Comment (mcompastie) [inline config/rm/nodesource/infrastructures-policies-mapping]: Some blank space to be removed, right ?

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 2]
```diff
@@ -670,6 +673,46 @@ public Topology getTopology(@HeaderParam("sessionid") String sessionId) throws N
...
+    @Path("infrastructuresmapping")
+    @Produces("application/json")
+    public Map<String, List<String>> getInfrasToPoliciesMapping(@HeaderParam("sessionid") String sessionId)
+            throws NotConnectedException {
+        ResourceManager rm = checkAccess(sessionId);
+        Map<String, List<String>> result = rm.getInfrasToPoliciesMapping();
+        Collection<PluginDescriptor> supportedInfrastructures = rm.getSupportedNodeSourceInfrastructures();
+        Collection<PluginDescriptor> supportedPolicies = rm.getSupportedNodeSourcePolicies();
+        Iterator<Map.Entry<String, List<String>>> iterator = result.entrySet().iterator();
+        while (iterator.hasNext()) {
```
Comment (mcompastie) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: Converting the result into a stream and using the filter() method can be a more concise method to consider here.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 3]
```diff
@@ -670,6 +673,46 @@ public Topology getTopology(@HeaderParam("sessionid") String sessionId) throws N
...
+            List<String> policies = entry.getValue().stream().filter(policy -> {
+                for (PluginDescriptor supportedPolicy : supportedPolicies) {
+                    if (supportedPolicy.getPluginName().equals(policy)) {
+                        return true;
+                    }
+                }
+                return false;
+            }).collect(Collectors.toList());
+            ;
+            return new AbstractMap.SimpleEntry<>(entry.getKey(), policies);
```
Comment (mcompastie) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: Please extract the content of the result.entrySet().stream().map() call to improve the readibility of your code. This is quiet obfuscated ! (this is a shared statement.)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java [hunk 4]
```diff
@@ -3003,6 +3007,40 @@ public void setNeededNodes(int neededNodes) {
...
+        Map<String, List<String>> mapping = new HashMap<>();
+        String fileName = null;
+        try {
+            fileName = PAResourceManagerProperties.RM_NODESOURCE_INFRA_POLICY_MAPPING.getValueAsString();
+            if (!(new File(fileName).isAbsolute())) {
+                // file path is relative, so we complete the path with the prefix RM_Home constant
+                fileName = PAResourceManagerProperties.RM_HOME.getValueAsString() + File.separator + fileName;
+            }
+
+            try (FileReader fr = new FileReader(fileName); BufferedReader br = new BufferedReader(fr)) {
```
Comment (mcompastie) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java]: Optionally, `Files.readAllLines()` provides a smoother way to read the definition file: I bet you can use its output directly as a stream.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java (hunk 3 above)
Comment (Mykhailenko) [reply rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: true

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java (hunk 2 above)
Comment (Mykhailenko) [reply rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: yeah, i dont know why I did not do it. good point

File: config/rm/nodesource/infrastructures-policies-mapping (hunk 1 above)
Comment (Mykhailenko) [reply config/rm/nodesource/infrastructures-policies-mapping]: yteah

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java (hunk 4 above)
Comment (Mykhailenko) [reply rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java]: Grazie mille

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java [hunk 5]
```diff
@@ -3003,6 +3006,37 @@ public void setNeededNodes(int neededNodes) {
...
+        Map<String, List<String>> mapping = new HashMap<>();
+        String fileName = null;
+        try {
+            fileName = PAResourceManagerProperties.RM_NODESOURCE_INFRA_POLICY_MAPPING.getValueAsString();
+            if (!(new File(fileName).isAbsolute())) {
+                // file path is relative, so we complete the path with the prefix RM_Home constant
+                fileName = PAResourceManagerProperties.RM_HOME.getValueAsString() + File.separator + fileName;
+            }
+
+            List<String> lines = Files.readAllLines(Paths.get(fileName));
```
Comment (mcompastie) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java]: (Optionnal) I was actually thinking about something alike:
```
Map<String,String[]> validLines = Files.readAllLines(Paths.get(fileName)).stream().map(line -> line.split(",")).filter(line -> (line.length() > 1)).toMap(splittedLine -> splittedLine.get(0).trim(), splittedLine.subList(1, strings.size());
```
(Approximately)
Therefore, you just have a fully prepared structure on which you can iterate to provision your mapping.

I'll leave this suggestion to you own consideration.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 6]
```diff
@@ -670,6 +673,47 @@ public Topology getTopology(@HeaderParam("sessionid") String sessionId) throws N
         return rm.getSupportedNodeSourceInfrastructures();
     }
 
+    @Override
+    @GET
+    @GZIP
+    @Path("infrastructuresmapping")
```
Comment (MaelAudren) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: I think that `/infrastructures/mapping` or `/infrastructures/policies` would be more relevant. It is better to keep the key "infrastructures".

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRestInterface.java [hunk 7]
```diff
@@ -551,6 +552,18 @@ boolean shutdown(@HeaderParam("sessionid") String sessionId,
...
+    /**
+     * @param sessionId a valid session
+     * @return a mapping which for each infrastructure provides list of policies
+     * that can work together with given infrastructure
+     */
+    @GET
+    @GZIP
+    @Path("infrastructuresmapping")
+    @Produces("application/json")
+    Map<String, List<String>> getInfrasToPoliciesMapping(@HeaderParam("sessionid") String sessionId)
```
Comment (MaelAudren) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRestInterface.java]: Same comment as upper

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java (hunk 6 above)
Comment (Mykhailenko) [reply rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: ok

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java (hunk 5 above)
Comment (Mykhailenko) [reply rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java]: Good thanks

**Pull request merged**